src/org/python/core/PyStringMap.java

```diff
@@ -240,25 +240,28 @@ final String stringmap_toString() {
         if (!ts.enterRepr(this)) {
             return "{...}";
         }
-        StringBuilder buf = new StringBuilder("{");
-        for (Entry<Object, PyObject> entry : table.entrySet()) {
-            Object key = entry.getKey();
-            if (key instanceof String) {
-                /* This is a bit complicated, but prevents us to duplicate
-                   PyString#__repr__ logic here. */
-                buf.append(new PyString((String)key).__repr__().toString());
-            } else {
-                buf.append(((PyObject)key).__repr__().toString());
+        StringBuilder buf = new StringBuilder( "{" );
+        try {
```

Comment:
Am I right in thinking the `try...finally` here is to protect the repr stack against bad implementations of `repr`? A thing we maybe should have had all along, rather than newly-necessary as part of this fix?

I'm not suggesting a separate PR, just trying to understand how it fits in.